gabi_test.go

```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+		})
+	}
+}
+
+func testNewKeyshareResponse(
+	t *testing.T,
+	keys map[string]*gabikeys.PublicKey, // the keys for which the keyshare protocol is done
+	ourSecret, userSecret *big.Int,
+	builders ProofBuilderList,
+	signature bool,
+) {
+	// The default value of this var is gabikeys.DefaultSystemParameters[2048].Lstatzk,
+	// but all test keys used here are 1024 bits, which use a smaller size for the secret key
+	// randomizer. So we lower this parameter because otherwise we exceed the maximum length of the
+	// secret key responses.
+	keyshareSecretRandomizerLength = gabikeys.DefaultSystemParameters[1024].Lstatzk
+
+	// Prepare some vars used throughout.
+	// We use keyNames below to keep track of the builders for which we need to do the keyshare
+	// protocol, as opposed to those for which we don't: if the index i of the i'th builder is in
+	// keyNames, we do the protocol, otherwise we don't.
+	keyNames := map[int]*string{}
+	var keysSlice []*gabikeys.PublicKey
+	for i, b := range builders {
+		keysSlice = append(keysSlice, b.PublicKey())
+		keyname := b.PublicKey().Issuer
+		if pk := keys[keyname]; pk != nil {
+			keyNames[i] = &pk.Issuer
+		}
+	}
+	nonce, err := GenerateNonce()
+	require.NoError(t, err)
+
+	// User chooses randomizer and computes h_W
+	userRandomizer, err := common.RandomBigInt(gabikeys.DefaultSystemParameters[1024].LmCommit)
+	require.NoError(t, err)
+	var hashInput []KeyshareChallengeInput[string]
+	for i, builder := range builders {
+		c, err := builder.Commit(map[string]*big.Int{"secretkey": userRandomizer})
+		require.NoError(t, err)
+		var otherComms []*big.Int
+		if len(c) > 2 {
+			otherComms = c[2:]
+		}
+		hashInput = append(hashInput, KeyshareChallengeInput[string]{
+			KeyID:            keyNames[i],
+			Value:            new(big.Int).Set(c[0]),
+			Commitment:       new(big.Int).Set(c[1]),
+			OtherCommitments: otherComms,
+		})
+	}
+	hashedComm, err := KeyshareUserCommitmentsHash(hashInput)
+	require.NoError(t, err)
+
+	// At this point, the user would send this to the KSS
+	req := KeyshareCommitmentRequest{HashedUserCommitments: hashedComm}
+
+	// The KSS responds with its commitments
+	ourRandomizer, ourComm, err := NewKeyshareCommitments(ourSecret, keysSlice)
+	require.NoError(t, err)
+
+	// User processes keyshare server commitment into the builders
+	for i, builder := range builders {
+		if keyNames[i] != nil {
+			builder.MergeProofPCommitment(ourComm[i])
+		}
+	}
+
+	// User computes challenge and user response
+	challenge, err := builders.ChallengeWithRandomizers(context, nonce, map[string]*big.Int{"secretkey": userRandomizer}, signature)
+	require.NoError(t, err)
+	userResponse := new(big.Int).Add(userRandomizer, new(big.Int).Mul(challenge, userSecret))
+
+	// User requests keyshare response
+	res := KeyshareResponseRequest[string]{
+		Nonce:              nonce,
+		UserResponse:       userResponse,
+		IsSignatureSession: signature,
+		ChallengeInput:     hashInput,
```

Comment:
I agree that `ChallengeInput` isn't a very good name, but I don't find `HashInput` very clarifying, because there is lots of hashing going on in all of Idemix.

How about `UserChallengeInput`? That makes it more clear that these parameters are the user's contribution to the input parameters of the challenge `c = H(...)`.